File: javalin/src/main/java/io/javalin/core/JavalinConfig.java
/*
 * Javalin - https://javalin.io
 * Copyright 2017 David Åse
 * Licensed under Apache 2.0: https://github.com/tipsy/javalin/blob/master/LICENSE
 */

package io.javalin.core;

import io.javalin.Javalin;
import io.javalin.core.compression.Brotli;
import io.javalin.core.compression.CompressionStrategy;
import io.javalin.core.compression.Gzip;
import io.javalin.core.plugin.Plugin;
import io.javalin.core.plugin.PluginAlreadyRegisteredException;
import io.javalin.core.plugin.PluginInitLifecycleViolationException;
import io.javalin.core.plugin.PluginLifecycleInit;
import io.javalin.core.plugin.PluginNotFoundException;
import io.javalin.core.security.AccessManager;
import io.javalin.core.security.SecurityUtil;
import io.javalin.core.util.CorsPlugin;
import io.javalin.core.util.Header;
import io.javalin.core.util.Headers;
import io.javalin.core.util.HeadersPlugin;
import io.javalin.core.util.HttpAllowedMethodsOnRoutesUtil;
import io.javalin.core.util.LogUtil;
import io.javalin.http.ContentType;
import io.javalin.http.ContextResolver;
import io.javalin.http.Handler;
import io.javalin.http.RequestLogger;
import io.javalin.http.SinglePageHandler;
import io.javalin.http.staticfiles.Location;
import io.javalin.http.staticfiles.ResourceHandler;
import io.javalin.http.staticfiles.StaticFileConfig;
import io.javalin.jetty.JettyResourceHandler;
import io.javalin.jetty.JettyUtil;
import io.javalin.plugin.json.JavalinJackson;
import io.javalin.plugin.json.JsonMapper;
import io.javalin.websocket.WsConfig;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Consumer;
import java.util.function.Supplier;
import java.util.stream.Stream;
import org.eclipse.jetty.server.Server;
import org.eclipse.jetty.server.session.SessionHandler;
import org.eclipse.jetty.servlet.ServletContextHandler;
import org.eclipse.jetty.websocket.server.JettyWebSocketServletFactory;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import static io.javalin.http.ContextResolverKt.CONTEXT_RESOLVER_KEY;
import static io.javalin.http.util.ContextUtil.maxRequestSizeKey;
import static io.javalin.plugin.json.JsonMapperKt.JSON_MAPPER_KEY;

public class JavalinConfig {
    // @formatter:off
    public boolean autogenerateEtags = false;
    public boolean prefer405over404 = false;
    public boolean enforceSsl = false;
    public boolean showJavalinBanner = true;
    @NotNull public String defaultContentType = ContentType.PLAIN;
    @NotNull public String contextPath = "/";
    public Long maxRequestSize = 1_000_000L; // either increase this or use inputstream to handle large requests
    @NotNull public Long asyncRequestTimeout = 0L;
    @NotNull public RoutingConfig routing = new RoutingConfig();
    @NotNull public Inner inner = new Inner();

    public static class RoutingConfig {
        public boolean ignoreTrailingSlashes = true;
        public boolean treatMultipleSlashesAsSingleSlash = false;
    }

    // it's not bad to access this, the main reason it's hidden
    // is to provide a cleaner API with dedicated setters
    public static class Inner {
        @NotNull public Map<Class<? extends Plugin>, Plugin> plugins = new LinkedHashMap<>();
        @NotNull public Map<String, Object> appAttributes = new HashMap<>();
        @Nullable public RequestLogger requestLogger = null;
        @Nullable public ResourceHandler resourceHandler = null;
        @NotNull public AccessManager accessManager = SecurityUtil::noopAccessManager;
        @NotNull public SinglePageHandler singlePageHandler = new SinglePageHandler();
        @Nullable public SessionHandler sessionHandler = null;
        @Nullable public Consumer<JettyWebSocketServletFactory> wsFactoryConfig = null;
        @Nullable public WsConfig wsLogger = null;
        @Nullable public Server server = null;
        @Nullable public Consumer<ServletContextHandler> servletContextHandlerConsumer = null;
        @NotNull public CompressionStrategy compressionStrategy = CompressionStrategy.GZIP;
    }
    // @formatter:on

    /**
     * Register a new plugin.
     */
    public void registerPlugin(@NotNull Plugin plugin) {
        if (inner.plugins.containsKey(plugin.getClass())) {
            throw new PluginAlreadyRegisteredException(plugin.getClass());
        }
        inner.plugins.put(plugin.getClass(), plugin);
    }

    /**
     * Get a registered plugin
     */
    public <T extends Plugin> T getPlugin(@NotNull Class<T> pluginClass) {
        T result = (T) inner.plugins.get(pluginClass);
        if (result == null) {
            throw new PluginNotFoundException(pluginClass);
        }
        return result;
    }

    public void enableDevLogging() {
        requestLogger(LogUtil::requestDevLogger);
        wsLogger(LogUtil::wsDevLogger);
        registerPlugin(new LogUtil.HandlerLoggingPlugin());
    }

    public void enableWebjars() {
        addStaticFiles(staticFiles -> {
            staticFiles.directory = "META-INF/resources/webjars";
            staticFiles.headers.put(Header.CACHE_CONTROL, "max-age=31622400");
        });
    }

    public void addStaticFiles(@NotNull String directory, @NotNull Location location) {
        addStaticFiles(staticFiles -> {
            staticFiles.directory = directory;
            staticFiles.location = location;
        });
    }

    public void addStaticFiles(@NotNull Consumer<StaticFileConfig> userConfig) {
        if (inner.resourceHandler == null) {
            inner.resourceHandler = new JettyResourceHandler();
        }
        StaticFileConfig finalConfig = new StaticFileConfig();
        userConfig.accept(finalConfig);
        inner.resourceHandler.addStaticFileConfig(finalConfig);
    }

    public void addSinglePageRoot(@NotNull String hostedPath, @NotNull String filePath) {
        addSinglePageRoot(hostedPath, filePath, Location.CLASSPATH);
    }

    public void addSinglePageRoot(@NotNull String hostedPath, @NotNull String filePath, @NotNull Location location) {
        inner.singlePageHandler.add(hostedPath, filePath, location);
    }

    public void addSinglePageHandler(@NotNull String hostedPath, @NotNull Handler customHandler) {
        inner.singlePageHandler.add(hostedPath, customHandler);
    }

    public void enableCorsForAllOrigins() {
        registerPlugin(CorsPlugin.forAllOrigins());
    }

    public void enableCorsForOrigin(@NotNull String... origins) {
        registerPlugin(CorsPlugin.forOrigins(origins));
    }

    public void accessManager(@NotNull AccessManager accessManager) {
        inner.accessManager = accessManager;
    }

    public void requestLogger(@NotNull RequestLogger requestLogger) {
        inner.requestLogger = requestLogger;
    }

    public void sessionHandler(@NotNull Supplier<SessionHandler> sessionHandlerSupplier) {
        inner.sessionHandler = sessionHandlerSupplier.get();
    }

    public void wsFactoryConfig(@NotNull Consumer<JettyWebSocketServletFactory> wsFactoryConfig) {
        inner.wsFactoryConfig = wsFactoryConfig;
    }

    public void wsLogger(@NotNull Consumer<WsConfig> ws) {
        WsConfig logger = new WsConfig();
        ws.accept(logger);
        inner.wsLogger = logger;
    }

    public void server(Supplier<Server> server) {
        inner.server = server.get();
    }

    public void configureServletContextHandler(Consumer<ServletContextHandler> consumer) {
        inner.servletContextHandlerConsumer = consumer;
    }

    public void compressionStrategy(Brotli brotli, Gzip gzip) {
        inner.compressionStrategy = new CompressionStrategy(brotli, gzip);
    }

    public void compressionStrategy(CompressionStrategy compressionStrategy) {
        inner.compressionStrategy = compressionStrategy;
    }

    public void globalHeaders(Supplier<Headers> headers) {
        registerPlugin(new HeadersPlugin(headers.get()));
    }

    public void jsonMapper(JsonMapper jsonMapper) {
        inner.appAttributes.put(JSON_MAPPER_KEY, jsonMapper);
    }

    public static void applyUserConfig(Javalin app, JavalinConfig config, Consumer<JavalinConfig> userConfig) {
        userConfig.accept(config); // apply user config to the default config

        AtomicBoolean anyHandlerAdded = new AtomicBoolean(false);
        app.events(listener -> {
            listener.handlerAdded(x -> anyHandlerAdded.set(true));
            listener.wsHandlerAdded(x -> anyHandlerAdded.set(true));
        });

        config.getPluginsExtending(PluginLifecycleInit.class)
            .forEach(plugin -> {
                plugin.init(app);
                if (anyHandlerAdded.get()) { // check if any "init" added a handler
                    throw new PluginInitLifecycleViolationException(((Plugin) plugin).getClass());
                }
            });

        config.inner.plugins.values().forEach(plugin -> plugin.apply(app));

        if (config.enforceSsl) {
            app.before(SecurityUtil::sslRedirect);
        }
        config.inner.appAttributes.putIfAbsent(JSON_MAPPER_KEY, new JavalinJackson());
        app.attribute(maxRequestSizeKey, config.maxRequestSize);

        config.inner.appAttributes.putIfAbsent(CONTEXT_RESOLVER_KEY, new ContextResolver());
    }

    private <T> Stream<? extends T> getPluginsExtending(Class<T> clazz) {
        return inner.plugins.values()
            .stream()
            .filter(clazz::isInstance)
            .map(plugin -> (T) plugin);
    }

    public void contextResolvers(@NotNull Consumer<ContextResolver> userResolver) {
        ContextResolver finalResolver = new ContextResolver();
        userResolver.accept(finalResolver);
        inner.appAttributes.put(CONTEXT_RESOLVER_KEY, finalResolver);
    }

    public void enableHttpAllowedMethodsOnRoutes() {
        registerPlugin(new HttpAllowedMethodsOnRoutesUtil());
    }
}

File: javalin/src/main/java/io/javalin/core/JavalinConfig.kt
/*
 * Javalin - https://javalin.io
 * Copyright 2017 David Åse
 * Licensed under Apache 2.0: https://github.com/tipsy/javalin/blob/master/LICENSE
 */
package io.javalin.core

import io.javalin.Javalin
import io.javalin.core.JettyUtil.disableJettyLogger
import io.javalin.core.compression.Brotli
import io.javalin.core.compression.CompressionStrategy
import io.javalin.core.compression.Gzip
import io.javalin.core.event.EventListener
import io.javalin.core.event.HandlerMetaInfo
import io.javalin.core.event.WsHandlerMetaInfo
import io.javalin.core.plugin.Plugin
import io.javalin.core.plugin.PluginAlreadyRegisteredException
import io.javalin.core.plugin.PluginInitLifecycleViolationException
import io.javalin.core.plugin.PluginLifecycleInit
import io.javalin.core.plugin.PluginNotFoundException
import io.javalin.core.security.AccessManager
import io.javalin.core.security.SecurityUtil
import io.javalin.core.util.CorsPlugin.Companion.forAllOrigins
import io.javalin.core.util.CorsPlugin.Companion.forOrigins
import io.javalin.core.util.LogUtil
import io.javalin.http.Context
import io.javalin.http.Handler
import io.javalin.http.RequestLogger
import io.javalin.http.SinglePageHandler
import io.javalin.http.staticfiles.JettyResourceHandler
import io.javalin.http.staticfiles.Location
import io.javalin.http.staticfiles.ResourceHandler
import io.javalin.http.staticfiles.StaticFileConfig
import io.javalin.websocket.WsConfig
import org.eclipse.jetty.server.Server
import org.eclipse.jetty.server.handler.ContextHandler.AliasCheck
import org.eclipse.jetty.server.session.SessionHandler
import org.eclipse.jetty.servlet.ServletContextHandler
import org.eclipse.jetty.websocket.servlet.WebSocketServletFactory
import java.util.concurrent.atomic.AtomicBoolean
import java.util.function.Consumer
import java.util.function.Supplier
import java.util.stream.Stream

/**
 * @JvmField will be fixed once Javalin.java is Javalin.kt
 */

class JavalinConfig {
    @JvmField var autogenerateEtags = false
    @JvmField var prefer405over404 = false
    @JvmField var enforceSsl = false
    @JvmField var precompressStaticFiles = false
    @JvmField var aliasCheckForStaticFiles: AliasCheck? = null
    @JvmField var showJavalinBanner = true
    @JvmField var logIfServerNotStarted = true
    @JvmField var ignoreTrailingSlashes = true
    @JvmField var defaultContentType = "text/plain"
    @JvmField var contextPath = "/"
    @JvmField var maxRequestSize = 1000000L // server will not accept payloads larger than 1mb by default
    @JvmField var asyncRequestTimeout = 0L
    @JvmField var inner: Inner = Inner()

    inner class Inner {
        @JvmField var plugins: MutableMap<Class<out Plugin>, Plugin> = mutableMapOf()
        @JvmField var appAttributes: MutableMap<Class<*>, Any> = mutableMapOf()
        @JvmField var requestLogger: RequestLogger? = null
        @JvmField var resourceHandler: ResourceHandler? = null
        @JvmField var accessManager: AccessManager = AccessManager { handler, ctx, roles -> SecurityUtil.noopAccessManager(handler, ctx, roles) }
        @JvmField var singlePageHandler = SinglePageHandler()
        @JvmField var sessionHandler: SessionHandler? = null
        @JvmField var wsFactoryConfig: Consumer<WebSocketServletFactory>? = null
        @JvmField var wsLogger: WsConfig? = null
        @JvmField var server: Server? = null
        @JvmField var servletContextHandlerConsumer: Consumer<ServletContextHandler>? = null
        @JvmField var compressionStrategy = CompressionStrategy.GZIP
    }

    fun registerPlugin(plugin: Plugin): JavalinConfig {
        if (inner.plugins.containsKey(plugin.javaClass)) {
            throw PluginAlreadyRegisteredException(plugin.javaClass)
        }
        inner.plugins[plugin.javaClass] = plugin
        return this
    }

    /**
     * Get a registered plugin
     */
    fun <T : Plugin> getPlugin(pluginClass: Class<T>): T {
        val plugin = inner.plugins[pluginClass] ?: throw PluginNotFoundException(pluginClass);
        return plugin as T;
    }

    fun enableDevLogging(): JavalinConfig {
        requestLogger(LogUtil::requestDevLogger);
        wsLogger(LogUtil::wsDevLogger);
        return this
    }

    fun enableWebjars(): JavalinConfig {
        return addStaticFiles("/webjars", Location.CLASSPATH)
    }

    fun addStaticFiles(classpathPath: String): JavalinConfig {
        return addStaticFiles(classpathPath, Location.CLASSPATH)
    }

    fun addStaticFiles(path: String, location: Location): JavalinConfig {
        return addStaticFiles("/", path, location)
    }

    fun addStaticFiles(urlPathPrefix: String, path: String, location: Location): JavalinConfig {
        disableJettyLogger()
        if (inner.resourceHandler == null) inner.resourceHandler = JettyResourceHandler(precompressStaticFiles, aliasCheckForStaticFiles)
        inner.resourceHandler!!.addStaticFileConfig(StaticFileConfig(urlPathPrefix, path, location))
        return this
    }

    fun addSinglePageRoot(path: String, filePath: String): JavalinConfig {
        addSinglePageRoot(path, filePath, Location.CLASSPATH)
        return this
    }

    fun addSinglePageRoot(path: String, filePath: String, location: Location): JavalinConfig {
        inner.singlePageHandler.add(path, filePath, location)
        return this
    }

    fun addSinglePageHandler(path: String, customHandler: Handler): JavalinConfig {
        inner.singlePageHandler.add(path, customHandler)
        return this
    }

    fun enableCorsForAllOrigins(): JavalinConfig {
        registerPlugin(forAllOrigins())
        return this
    }

    fun enableCorsForOrigin(vararg origins: String): JavalinConfig {
        registerPlugin(forOrigins(*origins))
        return this
    }

    fun accessManager(accessManager: AccessManager): JavalinConfig {
        inner.accessManager = accessManager
        return this
    }

    fun requestLogger(requestLogger: RequestLogger): JavalinConfig {
        inner.requestLogger = requestLogger
        return this
    }

    fun sessionHandler(sessionHandlerSupplier: Supplier<SessionHandler?>): JavalinConfig {
        disableJettyLogger()
        inner.sessionHandler = sessionHandlerSupplier.get()
        return this
    }

    fun wsFactoryConfig(wsFactoryConfig: Consumer<WebSocketServletFactory>): JavalinConfig {
        inner.wsFactoryConfig = wsFactoryConfig
        return this
    }

    fun wsLogger(ws: Consumer<WsConfig>): JavalinConfig {
        val logger = WsConfig()
        ws.accept(logger)
        inner.wsLogger = logger
        return this
    }

    fun server(server: Supplier<Server?>): JavalinConfig {
        inner.server = server.get()
        return this
    }

    fun configureServletContextHandler(consumer: Consumer<ServletContextHandler>?): JavalinConfig {
        inner.servletContextHandlerConsumer = consumer
        return this
    }

    fun compressionStrategy(brotli: Brotli?, gzip: Gzip?): JavalinConfig {
        inner.compressionStrategy = CompressionStrategy(brotli, gzip)
        return this
    }

    fun compressionStrategy(compressionStrategy: CompressionStrategy): JavalinConfig {
        inner.compressionStrategy = compressionStrategy
        return this
    }

    private fun <T> getPluginsExtending(clazz: Class<T>): Stream<out T> {
        return inner.plugins.values
                .stream()
                .filter { obj: Plugin? -> clazz.isInstance(obj) }
                .map { plugin: Plugin -> plugin as T }
    }

    companion object {
        fun applyUserConfig(app: Javalin, config: JavalinConfig, userConfig: Consumer<JavalinConfig?>) {
            userConfig.accept(config) // apply user config to the default config
            val anyHandlerAdded = AtomicBoolean(false)
            app.events { listener: EventListener ->
                listener.handlerAdded { x: HandlerMetaInfo? -> anyHandlerAdded.set(true) }
                listener.wsHandlerAdded { x: WsHandlerMetaInfo? -> anyHandlerAdded.set(true) }
            }
            config.getPluginsExtending(PluginLifecycleInit::class.java)
                    .forEach { plugin: PluginLifecycleInit ->
                        plugin.init(app)
                        if (anyHandlerAdded.get()) { // check if any "init" added a handler
                            throw PluginInitLifecycleViolationException((plugin as Plugin).javaClass)
                        }
                    }
            config.inner.plugins.values.forEach(Consumer { plugin: Plugin -> plugin.apply(app) })
            if (config.enforceSsl) {
                app.before(SecurityUtil::sslRedirect)
            }
        }
    }
}
